Step2 - 블랙잭

gmltkddl1 opened the pull request.

gmltkddl1 committed:
feat: step2 - 블랙잭

Readme 작성

gmltkddl1 committed:
feat: step2 - 블랙잭

요구사항 '카드는 에이스(1,11)부터 2~10 JACK(10) QUEEN(10) KING(10)이 있다' 구현

gmltkddl1 committed:
feat: step2 - 블랙잭

요구사항 '플레이어는 임의의 카드를 두장 지급 받는다 플레이어간의 카드는 중복되지 않는다' 구현

gmltkddl1 committed:
feat: step2 - 블랙잭

요구사항 '플레이어는 가진 카드의 합이 21이 넘지 않으면 추가 카드를 뽑을 수 있다' 구현

gmltkddl1 committed:
feat: step2 - 블랙잭

요구사항 '플레이어 중 카드의 수의 합이 21에 가까운 플레이어가 승리한다' 구현

gmltkddl1 committed:
feat: step2 - 블랙잭

controller 로직 구현

vsh123 commented on README.md:
```diff
@@ -1 +1,6 @@
-# kotlin-blackjack
\ No newline at end of file
+# kotlin-blackjack
```
기능 요구사항 작성 👍

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackCardsMap.kt:
```diff
@@ -0,0 +1,31 @@
+package blackjack.domain
+
+class BlackJackCardsMap(val cardsMap: MutableMap<BlackJackCard, Boolean> = mutableMapOf()) {
```
생각을 좀 바꿔본다면, 실제 블랙잭에서는 52장의 카드가 들어있는 덱에서 한장씩 카드를 뽑아주면서게임을 진행하게 되잖아요~?

그렇다면 최초 객체 생성시 52장의 카드가 들어있는 객체를만들고, 거기서 한장씩 꺼내도록 수정을 해보면 어떠띾요?

vsh123 commented on src/main/kotlin/blackjack/view/BlackJackView.kt:
```diff
@@ -0,0 +1,42 @@
+import blackjack.domain.BlackJackPlayers
+import blackjack.domain.BlackJackResult
+
+class BlackJackView {
```
[object class](https://kotlinlang.org/docs/object-declarations.html)를 사용해보면 어떨까요?

vsh123 commented on src/main/kotlin/blackjack/view/BlackJackView.kt:
```diff
@@ -0,0 +1,42 @@
+
+class BlackJackView {
+    companion object {
+        fun drawAndGetPlayerName(): List<String> {
```
Cotnroller →  View를 호출할때 draw도 필수일까요? `getPlayerName`  으로만 네이밍 해도 충분해 보여서요!

draw는 view의 선택으로 남기면 어떨까 싶습니다!

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackCards.kt:
```diff
@@ -0,0 +1,32 @@
+package blackjack.domain
+
+class BlackJackCards(val cards: MutableList<BlackJackCard>) {
```
mutlableList cards 프로퍼티가 외부로 노출되어서 외부에서 add가 가능한 것 같아요!
외부에서 add가 불가능하도록 수정해보면 어떨까요?

https://kotlinlang.org/docs/properties.html#backing-properties

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackCards.kt:
```diff
@@ -0,0 +1,32 @@
+
+    fun getCardsBestSum(): Int {
+        val cardValues = cards.map { it.number.values }
+
+        val sums =
+            cardValues.reduce { acc, list ->
+                acc.flatMap { accSum ->
+                    list.map { listSum -> accSum + listSum }
+                }
+            }
+        if (sums.any { it <= 21 }) {
+            return sums.filter { it <= 21 }.max()
+        }
```
21이라는 매직넘버가 많이 보이는 것 같아요!
이를 상수화해보면 어떨까요?

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackCards.kt:
```diff
@@ -0,0 +1,32 @@
+        cards.add(card)
+    }
+
+    fun isCardNumberSumUnder21(): Boolean {
```
21을 상수화한 이름으로 해당 함수도수정이 가능해보여요!

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackCards.kt:
```diff
@@ -0,0 +1,32 @@
+        val cardValues = cards.map { it.number.values }
+
+        val sums =
+            cardValues.reduce { acc, list ->
```
reduce와 fold의 차이에 대해 알아보고 어떤게 더 나을지 고민해보면 좋을 것 같습니다.

https://blog.leocat.kr/notes/2020/03/09/kotlin-reduce-and-fold

gmltkddl1 replied:
카드의 경우 빈값이 오는 경우가 없어서 reduce를 써도 무방해보이네요

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackCards.kt:
```diff
@@ -0,0 +1,32 @@
+        return cards.sumOf { it.number.values.min() } <= BlackJack_WIN_CARD_SUM
+    }
+
+    fun getCardsBestSum(): Int {
```
해당 클래스 및 함수에 대한 단위테스트가 보이질 않는 것 같아요~ 테스트코드를 추가해보면 어떨까요?

vsh123 commented on src/test/kotlin/blackjack/BlackJackCardTest.kt:
```diff
@@ -0,0 +1,25 @@
+import org.assertj.core.api.Assertions.assertThatThrownBy
+import org.junit.jupiter.api.Test
+
+class BlackJackCardTest {
```
테스트 패키지의 위치가 도메인 클래스와 다른 것 같아요! 패키지 위치를맞춰주면 어떨까요?

vsh123 commented on src/main/kotlin/blackjack/Main.kt:
```diff
@@ -0,0 +1,36 @@
+
+    blackJackPlayers.players.forEach {
+        while (it.isDrawPossible()) {
+            val yn = BlackJackView.drawAndGetPlayerDrawCardYn(it)
+            if (yn.equals("n")) {
```
View에서 Boolean값을 전달 받도록 로직을 고쳐보면 어떨까요??

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackCards.kt:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    fun isCardNumberSumUnder21(): Boolean {
+        return cards.sumOf { it.number.values.min() } <= BlackJack_WIN_CARD_SUM
```
`getCardsBestSum()` 이 아닌 `min()`으로만 계산하신 이유가 있을까요? 결과가 다를거 같아요!

gmltkddl1 replied:
해당 조건은 추가 카드 획득 조건이고 bestSum과 최솟값을 이용하는 경우 결국 결과는 같다고 생각합니다.
결국 내손의 있는 카드의 value 조합 중 21을 넘지 않는 조합이 한개라도 있으면 카드를 추가로 뽑을 수 있어야하고
그중 가장 작은 것이 최솟값의 합이기 때문입니다.

vsh123 reviewed (changes requested):
안녕하세요~!

step2 미션 구현 잘해주셨네요!

몇가지 코멘트를 남겼으니 확인부탁드립니다!

gmltkddl1 committed:
refactor: step2 - 블랙잭

view에서 도메인 조직 제거
테스트 클래스 패키지 구조 실제 프로젝트와 동일하게 변경
BlackJackCards 테스트 코드 추가
매직넘버 제거
backing property 추가
controller의 view 호출 함수 이름 변경

gmltkddl1 committed:
refactor: step2 - 블랙잭

카드 뽑기 중복을 제거하는 방식 대신 리스트 생성해서 꺼내 쓰도록 수정

vsh123 commented on src/test/kotlin/blackjack/domain/BlackJackCardMapTest.kt:
```diff
@@ -0,0 +1,33 @@
+import org.assertj.core.api.Assertions.assertThatThrownBy
+import org.junit.jupiter.api.Test
+
+class BlackJackCardMapTest {
```
도메인 객체와 테스트명을 맞춰주는게 어떨까요! 

```suggestion
class BlackJackCardsMapTest {
```

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackCardsMap.kt:
```diff
@@ -0,0 +1,28 @@
+package blackjack.domain
+
+class BlackJackCardsMap(
```
현재 해당 프로퍼티는 Map을 들고있지 않아서, `~Map` 이라고 네이밍할 이유가 없어보여요!! 

[네이밍](https://tecoble.techcourse.co.kr/post/2020-04-24-variable_naming/)을 읽어보고 더 명확한 네이밍을 고려해보면 어떨까요?

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackCardsMap.kt:
```diff
@@ -0,0 +1,28 @@
+
+class BlackJackCardsMap(
+    private val cards: MutableList<BlackJackCard> = mutableListOf(),
+    blackJackCardsSetter: BlackJackCardsSetter = RandomBlackJackCardsSetter(),
```
init에서 cards에 add를 해주면서 52장을 넣어주는데요
이렇게 되면 다음과 같은 고민을 해볼 수 있을 것 같아요.

* 만약 BlackJackCardsMap에 원하는 카드들만 넣어서 테스트를 작성하고 싶을때는 어떻게하면 좋을까?
  * 예를들어 A한장, K한장만 넣어서 테스트를 하고싶은데 지금은 2장 + 52장이 셋팅 될 것 같아요.

이걸 위해 BlackJackCardsSetter라는걸 추가하신걸로 보이는데,  blackJackCardsSetter는 결국 다른 객체가 `BlackJackCardsMap`의 cards를 핸들링한다는 점에서 셋팅된 프로퍼티를 핸들링할 수 있는 기능을 외부로 노출하는 것과 다르지 않다는 생각이 드네요!

순서를 조금 바꾸어서, 52장의 카드를 만들어서 객체를 생성해보면 어떨까요?
대표적으로는 팩터리 메서드가 생각나네요!! 이렇게하면 `BlackJackCardsSetter` 도 없앨 수 있을 것 같아요!

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackCardsMap.kt:
```diff
@@ -0,0 +1,28 @@
+        blackJackCardsSetter.shuffle(cards)
+    }
+
+    fun get(): BlackJackCard {
```
블랙잭에 맞게끔 `draw`같은 네이밍은 어떠씬가요?

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackCardsMap.kt:
```diff
@@ -0,0 +1,28 @@
+
+    fun get(): BlackJackCard {
+        require(cards.isNotEmpty()) { "카드가 없어요" }
+        val blackJackCard = cards.first()
```
first와 removeFirst도 괜찮지만, List말고 다른 구조형을 적용해보면 어떨까요? Stack이나 Queue등이 생각나넨요!

vsh123 commented on src/test/kotlin/blackjack/domain/BlackJackCardMapTest.kt:
```diff
@@ -0,0 +1,33 @@
+            }
+        val blackJackCard = BlackJackCardsMap(blackJackCardsSetter = blackJackCardsSetter).get()
+        assertThat(blackJackCard.number.values).containsExactly(1, 11)
+        assertThat(blackJackCard.shape.shape).isEqualTo("HEART")
```
아래와 같이 equals를 테스트하면, BlackJackCardShape에 shape프로퍼티가 필요없어보이네요! 

```suggestion
        assertThat(blackJackCard.shape).isEqualTo(BlackJackCardShape.HEART)
```

vsh123 commented on src/main/kotlin/blackjack/Main.kt:
```diff
@@ -0,0 +1,35 @@
+
+    blackJackPlayers.players.forEach {
+        while (it.isDrawPossible()) {
+            if (!BlackJackView.getPlayerDrawCardYn(it)) {
```
> indent(인덴트, 들여쓰기) depth를 2를 넘지 않도록 구현한다. 1까지만 허용한다.

라는 규칙에 부합하지 못한것 같아요!
규칙에 맞게끔 수정을 한번 진행해보면 어떨까요?

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackCards.kt:
```diff
@@ -0,0 +1,38 @@
+class BlackJackCards(private val _cards: MutableList<BlackJackCard> = mutableListOf()) {
+    val cards: List<BlackJackCard>
+        get() {
+            return _cards.toList()
```
backing properties활용 👍

vsh123 commented on src/main/kotlin/blackjack/domain/BlackJackCards.kt:
```diff
@@ -0,0 +1,38 @@
+
+    fun isCardNumberSumUnderBlackJackWinCardSum(): Boolean {
+        val bestSum = getCardsBestSum()
+        return (bestSum in 1..BLACKJACK_WIN_CARD_SUM)
```
bestSum이 0이하가 나올 수 없고, 숫자 비교니까 아래처럼해도 괜찮지 않을까요? 

```suggestion
        return bestSum <= BLACKJACK_WIN_CARD_SUM
```

gmltkddl1 replied:
22가 넘는 경우 -1를 리턴하도록 구현하였기 때문에 양수 처리가 필요합니다

vsh123 reviewed (changes requested):
안녕하세요!
피드백 반영 빠르게 잘해주셨네요 👍 

몇가지 코멘트를 남겨드렸는데요! 확인 부탁드리겠습니다 :)

gmltkddl1 committed:
refactor: step2 - 블랙잭

클래스 메소드 네이밍 변경
BlackJackCards factory method 추가에서 카드 초기화하도록 수정
BlackJackCards java stack으로 변경
BlackJackCardShape의 불필요한 property 삭제
indent가 2넘는 코드 수정

gmltkddl1 committed:
refactor: step2 - 블랙잭

테스트 coverage 개선

vsh123 reviewed (approved):
안녕하세요!

step2 반영 잘해주셨네요 💯 
마무리 코멘트는 온라인 리뷰로 갈음하도록 하겠습니다.

다음스텝에서 뵐게요!!
